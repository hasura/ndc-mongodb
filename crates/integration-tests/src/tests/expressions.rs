use insta::assert_yaml_snapshot;
use ndc_models::ExistsInCollection;
use ndc_test_helpers::{
    asc, binop, exists, field, query, query_request, relation_field, relationship, target, value,
};

use crate::{connector::Connector, graphql_query, run_connector_query};

#[tokio::test]
async fn evaluates_field_name_that_requires_escaping_in_nested_expression() -> anyhow::Result<()> {
    // Skip this test in MongoDB 5 because the example fails there. We're getting an error:
    //
    // > Kind: Command failed: Error code 5491300 (Location5491300): $documents' is not allowed in user requests, labels: {}
    //
    // This means that remote joins are not working in MongoDB 5
    if let Ok(image) = std::env::var("MONGODB_IMAGE") {
        if image == "mongo:5" {
            return Ok(());
        }
    }

    assert_yaml_snapshot!(
        graphql_query(
            r#"
                query Filtering {
                  extendedJsonTestData(where: { value: { _regex: "hello" } }) {
                    type
                    value
                  }
                }
            "#
        )
        .run()
        .await?
    );
    Ok(())
}

#[tokio::test]
async fn evaluates_exists_with_predicate() -> anyhow::Result<()> {
    assert_yaml_snapshot!(
        run_connector_query(
            Connector::Chinook,
            query_request()
                .collection("Artist")
                .query(
                    query()
                        .predicate(exists(
                            ExistsInCollection::Related {
                                relationship: "albums".into(),
                                arguments: Default::default(),
                            },
                            binop("_iregex", target!("Title"), value!("Wild"))
                        ))
                        .fields([
                            field!("_id"),
                            field!("Name"),
                            relation_field!("albums" => "albums", query().fields([
                                field!("Title")
                            ]).order_by([asc!("Title")]))
                        ]),
                )
                .relationships([("albums", relationship("Album", [("ArtistId", "ArtistId")]))])
        )
        .await?
    );
    Ok(())
}
