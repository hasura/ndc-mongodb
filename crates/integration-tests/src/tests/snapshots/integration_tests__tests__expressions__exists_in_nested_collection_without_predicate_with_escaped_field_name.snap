---
source: crates/integration-tests/src/tests/expressions.rs
expression: "run_connector_query(Connector::TestCases,\n            query_request().collection(\"weird_field_names\").query(query().predicate(Expression::Exists {\n                                in_collection: ExistsInCollection::NestedCollection {\n                                    column_name: \"$invalid.array\".into(),\n                                    arguments: Default::default(),\n                                    field_path: Default::default(),\n                                },\n                                predicate: None,\n                            }).fields([field!(\"_id\"),\n                                field!(\"invalid_array\" => \"$invalid.array\",\n                                    array!(object!([field!(\"invalid_element\" =>\n                                    \"$invalid.element\")])))]).order_by([asc!(\"$invalid.name\")]))).await?"
---
- rows:
    - _id: 66cf91a0ec1dfb55954378bd
      invalid_array:
        - invalid_element: 1
    - _id: 66cf9230ec1dfb55954378be
      invalid_array:
        - invalid_element: 2
    - _id: 66cf9274ec1dfb55954378bf
      invalid_array:
        - invalid_element: 3
    - _id: 66cf9295ec1dfb55954378c0
      invalid_array:
        - invalid_element: 4
